presto-main/src/test/java/io/prestosql/sql/planner/TestDomainCoercer.java

```diff
@@ -0,0 +1,166 @@
+{
+    private Metadata metadata;
+    private TypeOperators typeOperators;
+
+    @BeforeClass
+    public void setup()
+    {
+        metadata = createTestMetadataManager();
+        typeOperators = new TypeOperators();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        metadata = null;
+    }
+
+    private Domain applySaturatedCasts(Domain domain, Type coercedValueType)
+    {
+        return DomainCoercer.applySaturatedCasts(metadata, typeOperators, TEST_SESSION, domain, coercedValueType);
+    }
+
+    @Test
+    public void testNone()
+    {
+        assertEquals(applySaturatedCasts(Domain.none(BIGINT), INTEGER), Domain.none(INTEGER));
+    }
+
+    @Test
+    public void testAll()
+    {
+        assertEquals(applySaturatedCasts(Domain.all(BIGINT), INTEGER), Domain.all(INTEGER));
+    }
+
+    @Test
+    public void testOnlyNull()
+    {
+        assertEquals(applySaturatedCasts(Domain.onlyNull(BIGINT), INTEGER), Domain.onlyNull(INTEGER));
+    }
+
+    @Test
+    public void testCoercedValueSameAsOriginal()
+    {
+        assertEquals(
+                applySaturatedCasts(multipleValues(BIGINT, ImmutableList.of(1L, 10000L, -2000L)), SMALLINT),
+                multipleValues(SMALLINT, ImmutableList.of(1L, 10000L, -2000L)));
+
+        Domain original = Domain.create(
+                ValueSet.ofRanges(
+                        lessThan(DOUBLE, 0.0),
+                        range(DOUBLE, 0.0, false, 1.0, false),
+                        range(DOUBLE, 2.0, true, 3.0, true),
+                        greaterThan(DOUBLE, 4.0)),
+                true);
+        assertEquals(
+                applySaturatedCasts(original, REAL),
+                Domain.create(
+                        ValueSet.ofRanges(
+                                lessThan(REAL, (long) Float.floatToIntBits(0.0f)),
+                                range(REAL, (long) Float.floatToIntBits(0.0f), false, (long) Float.floatToIntBits(1.0f), false),
+                                range(REAL, (long) Float.floatToIntBits(2.0f), true, (long) Float.floatToIntBits(3.0f), true),
+                                greaterThan(REAL, (long) Float.floatToIntBits(4.0f))),
+                        true));
+    }
+
+    @Test
+    public void testOutsideTargetTypeRange()
+    {
+        assertEquals(
+                applySaturatedCasts(
+                        multipleValues(BIGINT, ImmutableList.of(1L, 10000000000L, -2000L)), SMALLINT),
+                multipleValues(SMALLINT, ImmutableList.of(1L, -2000L)));
+
+        assertEquals(
+                applySaturatedCasts(
+                        Domain.create(
+                                ValueSet.ofRanges(range(DOUBLE, 0.0, true, ((double) Float.MAX_VALUE) * 10, true)),
+                                true),
+                        REAL),
+                Domain.create(
+                        ValueSet.ofRanges((greaterThanOrEqual(REAL, (long) Float.floatToIntBits(0.0f)))),
```

Comment:
add a test case where low is outside of target type range.

Reply:
The next test case has such a scenario 
`range(DOUBLE, ((double) Float.MAX_VALUE) * -2, true, ((double) Float.MAX_VALUE) * 10, true)`
Did you mean a test where only low is outside target type range ?